lib/spack/spack/directives.py

```diff
@@ -598,6 +600,23 @@ def _execute_depends_on(pkg: "spack.package_base.PackageBase"):
     return _execute_depends_on
 
 
+@directive(dicts=())
+def redistribute(source=True, binary=True):
+    """Can be used inside a Package definition to declare that
+    the package source and/or compiled binaries should not be
+    redistributed.
+
+    By default, Packages allow source/binary distribution (i.e. in
+    mirrors).
+    """
+
+    def _execute_redistribute(pkg: "spack.package_base.PackageBase"):
+        pkg.redistribute_source = source
+        pkg.redistribute_binary = binary
```

Comment:
I think these should be stored in a when-keyed dictionary, e.g.:

```python
@directive("redistribute")
def redistribute(source=True, binary=True, when: WhenType = None):
    when_spec = _make_when_spec(when)
    if not when_spec:
        return

    redist = self.redistribute.setdefault(when_spec, {})
    redist["source"] = source
    redist["binary"] = binary
```

note that this doesn't affect the solve -- we should just use it after the solve to determine if a concrete spec is redistributable.

I actually *do* think something like this:

```python
redistribute: {
    "source": true,
    "binary": true,
}
```

should go on the spec itself, but I can add that part in #32312 (where I think it makes more sense)

Reply:
> note that this doesn't affect the solve -- we should just use it after the solve to determine if a concrete spec is redistributable.

Currently there are spots where we don't want to concretize specs (e.g. when we do `spack mirror --all`). My concern with any `when=` approach is that - unless the `when` is strictly limited to versions - not all information will be available at mirror creation time.

We could also detect conditional redistribution settings, conservatively return `False` for non-concrete specs (thereby sacrificing the possibility of including them in a full mirror) and allowing a more fine-grained check when the associated Spec is concrete.

Reply:
> My concern with any when= approach is that - unless the when is strictly limited to versions - not all information will be available at mirror creation time.

Yeah this is a good point.  I may be thinking about this too quickly, but I would say that *if* the user provides the `source=` argument, you could require the `when` spec to be either `True`, `False`, or a constraint only on version.  e.g., this would be ok:

```python
    redistribute(source=False, binary=False, when="@:3")
    redistribute(binary=False, when="+option")
```

but not this, for the reason you described:
```python
    redistribute(source=False, binary=False, when="@:3")
    redistribute(source=False, binary=False, when="+option")
```

I agree -- almost all we know at source mirror generation time the version of the tarball. 

It occurs to me that we currently don't have any redistribution information for specific patches (which can be conditional on a variant, which means you might know that *they* shouldn't be redistributed when something other than a version is constrained), but I also think every patch is at least somewhat dependent on a version, and we should leave patches for some future PR 😅.

Reply:
Regarding `@directive("redistribute")`

I think it would be best to use `@directive(dicts=())` because I don't think the following makes sense:

```
redistribute(source=False, when=@2:4)
redistribute(source=True, when=@3.1)
```

I don't think we should have to offer semantics for competing enablement/disablement of redistribution, and for that reason I think users should only allowed to be set `source/binary=False` (never True); as a consequence, there is no need for a dict to store this information (all we need is a list of when specs that disable source/binary distribution).

In any case, the part about accepting a `when_spec` is handled (but not yet tested, so for now I have marked this PR as a draft).

Reply:
@scheibelp I want the dictionaries to be consistent and indexed by `when` spec so that we can implement consistent logic in directives like `drop()` (see #35045). Before #40326, we had a bunch of different directive dictionary structures, which made it very hard to understand which dictionaries were structured which way.  It's better to be consistent on this than to have to implement different algorithms for filtering different dicts.

If you want to simplify the above, one option would be to do what the `license()` directive already does, which is to detect when ranges are overlapping and disallow it. Maybe generalizing [the logic here](https://github.com/spack/spack/pull/39346/files#diff-5a5ca8628a0ed62863ebe23f74fa01a6152c018d7b6693cea4f57857db22ed62R872-R885) makes sense?

Reply:
> I want the dictionaries to be consistent and indexed by when spec

OK: that is done now, it looked like the values can be anything, so I made a new `DisableRedistribute` tuple

> If you want to simplify the above, one option would be to do what the license() directive already does, which is to detect when ranges are overlapping and disallow it. 

For enable/disable conflicts, these are resolved by restricting the user to disable. For disable/disable conflicts, I'm not sure if we should actually consider them conflicts in general:

```
class TestPackage():
  redistribute(source=False, when="@1.1")
  redistribute(binary=False, when="@1.1")
  redistribute(binary=False, when=@1.0:1.2+foo)
```

the first two items can conflict, but I think it's worth allowing the 2nd and 3rd together.

If you agree, then the conflict checking is just key collision.